File: python/testcigar.py
import json
from jsonargparse import CLI

import jax.numpy as jnp

import cigartree
import likelihood
import h20

def main (treeFilename: str,
          alignFilename: str,
          modelFilename: str,
          discretize: bool = False,
          ):
    """
    Compute the log likelihood of a tree given an alignment and a model.
    
    Args:
        treeFilename: Newick tree file
        alignFilename: FASTA alignment file
        modelFilename: Historian-format JSON file with model parameters
        discretize: Discretize branch lengths
    """
    with open(treeFilename, 'r') as f:
        treeStr = f.read()
    with open(alignFilename, 'r') as f:
        alignStr = f.read()
    with open(modelFilename, 'r') as f:
        modelJson = json.load (f)

    ct = cigartree.makeCigarTree (treeStr, alignStr)

    alphabet, mixture, indelParams = likelihood.parseHistorianParams (modelJson)
    seqs, nodeName, distanceToParent, parentIndex, transCounts = cigartree.getHMMSummaries (treeStr, alignStr, alphabet)

    subRate, rootProb = mixture[0]
    if discretize:
        discSubMatrix = likelihood.computeSubMatrixForDiscretizedTimes (subRate)
        subMatrix = likelihood.computeSubMatrixForDiscretizedBranchLengths (distanceToParent, discSubMatrix)
        subll = likelihood.subLogLikeForMatrices (seqs, parentIndex, subMatrix, rootProb)
    else:
        subll = likelihood.subLogLike (seqs, distanceToParent, parentIndex, subRate, rootProb)
    subll_total = float (jnp.sum (subll))

    transMat = jnp.stack ([h20.dummyRootTransitionMatrix()] + [h20.transitionMatrix(t,indelParams,alphabetSize=len(alphabet)) for t in distanceToParent[1:]], axis=0)
    transMat = jnp.log (jnp.maximum (transMat, h20.smallest_float32))
    transll = transCounts * transMat
    transll_total = float (jnp.sum (transll))

    print (json.dumps({'loglike':{'subs':subll_total,'indels':transll_total}, 'cigartree': ct}))

if __name__ == '__main__':
    CLI(main, parser_mode='jsonnet')

File: js/testcigar.js

import fs from 'fs';
import { makeCigarTree, expandCigarTree, countGapSizes } from './cigartree.js';
import { parseHistorianParams, subLogLike, transLogLike, sum } from './likelihood.js';

if (process.argv.length != 5) {
    console.error('Usage: ' + process.argv[1] + ' model.json tree.nh align.fa');
    process.exit(1);
}

const [ modelFilename, treeFilename, alignFilename ] = process.argv.slice(2);
const modelJson = JSON.parse (fs.readFileSync(modelFilename).toString());
const treeStr = fs.readFileSync(treeFilename).toString();
const alignStr = fs.readFileSync(alignFilename).toString();

const ct = makeCigarTree (treeStr, alignStr);
const { alignment, expandedCigar, distanceToParent, leavesByColumn, internalsByColumn, branchesByColumn } = expandCigarTree (ct);
const lcAlignment = alignment.map ((s) => s.toLowerCase());

const { alphabet, mixture, indelParams } = parseHistorianParams (modelJson);
const { transCounts } = countGapSizes (expandedCigar);

const { subRate, rootProb } = mixture[0];
const subll = subLogLike (lcAlignment, distanceToParent, leavesByColumn, internalsByColumn, branchesByColumn, alphabet, rootProb, { subRate });
const subll_total = sum (subll);

const transll = transLogLike (transCounts, distanceToParent, { indelParams, alphabet });
const transll_total = sum (transll);

console.log (JSON.stringify({'loglike':{'subs':subll_total,'indels':transll_total}, 'cigartree': ct}));

